### DISCUSSION TITLE
Forcing density in cst & Priority optimisation, and dose statistics

### DISCUSSION THREAD

**Discussion by Fabio-Dan** [e0404/matRad#707] [Q&A]
Hi all,

I've been using the MatRad dev_varRBErobOpt fork for some treatment planning studies and have a few questions - apologies in advance for the length!

Is there a relatively simple way to change the underlying HU of a CT for a given contour found in the CST (a contour containing contrast, for example, and forcing that to a water density similar to commercial TPS)? Or am I left to changing these HU values in another software suite (Slicer, etc.)? I tried to follow the example where geometric shapes are generated and HU values are assigned to a given contour, but I didn't have much luck doing this with existing contours contained in the CST.

I am using datasets which have multiple PTV dose levels (usually contained within each other) - for example, a Prostate plan with PTVs @ 60Gy + 40Gy. These plans also contain rinds (donut-like shapes) to control dose fall-off and subtracted Low-Dose PTV volumes (PTV60 minus PTV40, for example) to derive dose statistics for the lower dose level. MatRad lacks the capability to handle both external and internal contours on the same slice, thus filling in these areas.
![image](https://github.com/e0404/matRad/assets/104505852/20cb3667-c86b-45df-8414-424c95da56d5)

To check my understanding of the optimiser: Can I achieve the required dose fall-off by using the priority mechanism? In the example above, there are two PTVs - PTV1 and a larger PTV2 that envelops the smaller PTV1. If I make PTV1 Priority 1 and PTV2 Priority 2, will the objectives in PTV1 (60Gy) override those in PTV2 (40Gy) which overlap the same voxels, and in turn, will the objectives associated with PTV40 'only' function on the voxels outside of PTV 60?

As for the dose statistics, I'm guessing there is no easy way to add a channel like on some commercial TPS's on each slice to force the 'donut' subtraction - if I was only interested in the dose stats for PTV40 minus PTV60, is there another way to achieve this?
![image](https://github.com/e0404/matRad/assets/104505852/b7de7383-dece-453e-aa7e-cf889ef77b5c)

I currently use the 'matRad_calcDVH(cst,resultGUI.physicalDose)' and then pull out the required dose stats for example:

    dvh1 = matRad_calcDVH(cst,resultGUI.physicalDose);
    for j = 1:numStructures
        dose = dvh1(j).doseGrid;
        volume = dvh1(j).volumePoints;
        [~, d2_idx] = min(abs(volume - 2));
        d2(j) = dose(d2_idx);
        
        ect.... 

Thanks for your help!

**Comment by wahln** [comment]
Thanks for your interest in matRad, let me give you some pointers:
- HU overrides are possible using the data in the 4th column of the `cst`. There's no specific function for that, but since, for example, cst{i,4}{1} contains the index list of voxels inside the i-th structure, you can use it to override HU: ct.cubeHU{1}(cst{i,4}{1}) = 0. There are also ways to to this directly on rED / rSP before dose calculation, but this requires a bit more hacking. If you are missing this functionality as a feature, please write a separate concise issue to request that feature. We are currently not touching the dose calculation, as we do a substantial refactoring of it, so development of new features is a little bit slow on that front.
- "MatRad lacks the capability to handle both external and internal contours on the same slice, thus filling in these areas." I don't know what you mean by that, could you explain in a bit more detail?
- Overlapping structures are handled and cut with the overlap priorities during optimization if priorities differ, but in dose stats evaluation it uses the full volume. Playing around with the overlap priorities in the cst and manually using the function `cstOverlap = matRad_setOverlapPriorities(cst)` can get you quite far by creating a cstOverlap where overlapping volumes are stripped/cut, which can then be passed into the evaluation functions for DVH / QI calculations to only calculate on the overlapped volume.
- In all this stuff one needs to be a bit careful with the resolution of the dose grid and the corresponding interpolation / downsampling going on, particularly if one interferes directly with dose calculation / optimization. 

Let me know if this sounds like it could be useful for you and where you would need more detail. As many open source software tools, matRad does not have that many quality-of-life functionalities as commercial tps
